### PR TITLE
Modify the motd command to only send to unit users

### PR DESCRIPTION
**Pull request by rv0lt**
## Read this before submitting the PR

1. Always create a Draft PR first
2. Go through sections 1-5 below, fill them in and check all the boxes
3. Make sure that the branch is updated; if there's an "Update branch" button at the bottom of the PR, rebase or update branch.
4. When all boxes are checked, information is filled in, and the branch is updated: mark as Ready For Review and tag reviewers (top right)
5. Once there is a submitted review, implement the suggestions (if reasonable, otherwise discuss) and request an new review.

If there is a field which you are unsure about, enter the edit mode of this description or go to the [PR template](../.github/pull_request_template.md); There are invisible comments providing descriptions which may be of help.

## 1. Description / Summary

TODO

## 2. Jira task / GitHub issue

[_Link to the github issue or add the Jira task ID here._](https://scilifelab.atlassian.net/jira/software/projects/DDS/boards/13?selectedIssue=DDS-2105)

## 3. Type of change

What _type of change(s)_ does the PR contain?

**Check the relevant boxes below. For an explanation of the different sections, enter edit mode of this PR description template.**

- [X] New feature
  - [ ] Breaking: _Why / How? Add info here._ 
  - [X] Non-breaking 
- [ ] Database change: _Remember the to include a new migration version, **or** explain here why it's not needed._ 
- [ ] Bug fix 
- [ ] Security Alert fix 
  - [ ] Package update 
    - [ ] Major version update 
- [ ] Documentation 
- [ ] Workflow 
- [ ] Tests **only** 

## 4. Additional information

- [ ] [Sprintlog](../SPRINTLOG.md) 
- [ ] Blocking PRs 
  - [ ] Merged 
- [ ] PR to `master` branch: \_If checked, read [the release instructions](../doc/procedures/new_release.md) 
  - [ ] I have followed steps 1-8. 

## 5. Actions / Scans

_Check the boxes when the specified checks have passed._

**For information on what the different checks do and how to fix it if they're failing, enter edit mode of this description or go to the [PR template](../.github/pull_request_template.md).**

- [ ] **Black**

- [ ] **Prettier**

- [ ] **Yamllint**

- [ ] **Tests**

- [ ] **CodeQL**

- [ ] **Trivy**

- [ ] **Snyk**

### CODE REVIEW

**Review by valyo — CHANGES_REQUESTED** [review]
Apart from that particular suggestion, this change cannot be tested (and used, once merged) without a corresponding change in CLI.

**Comment by rv0lt** [comment]
> Apart from that particular suggestion, this change cannot be tested (and used, once merged) without a corresponding change in CLI. Add sprintlog

True about the client, but there is a way to test it locally. By calling the endpoint directly, from your terminal.
Login in first with the superadmin account in the local instance of dds running in localhost and then:

curl -X 'POST' \
  'localhost:5000/api/v3/motd/send' \
  -H 'accept: application/json' \
  -H 'X-CLI-Version: 3.0.1' \
  -H 'Authorization: Bearer TOKEN_SAVED_IN_DDS_CLI_FILE' \
  -H 'Content-Type: application/json' \
  -d '{
  "motd_id": 2,
  "unit_personnel_only": true
    }'
    

Just need to change the TOKEN_SAVED_IN_DDS_CLI_FILE with the real token. You can also get the syntax via the swagger UI in http://localhost:5000/api/documentation/v3/#/

File: SPRINTLOG.md [hunk 1]
```diff
@@ -431,3 +431,7 @@ _Nothing merged during this sprint_
 - Flask command to update unit quotas ([#1551](https://github.com/ScilifelabDataCentre/dds_web/pull/1551))
 - Bump python base image to 3.12 and related libraries in both web and client([#1548](https://github.com/ScilifelabDataCentre/dds_web/pull/1548))
 - Warning_level option defaults to 0.8([#1557](https://github.com/ScilifelabDataCentre/dds_web/pull/1557))
+
+# 2024-09-24 - 2024-10-04
+
+- Modify the motd command to only send to unit users([#1552](https://github.com/ScilifelabDataCentre/dds_web/pull/1552))
```
Comment (valyo) [inline SPRINTLOG.md:437]:
```suggestion
- Add option to motd command for sending to unit users only([#1552](https://github.com/ScilifelabDataCentre/dds_web/pull/1552))
```

**Review by valyo — CHANGES_REQUESTED** [review]

**Review by valyo — APPROVED** [review]
Everything seems good to me. Probably don't yet merge though, before Ina's approval. I think she wanted to explicitly approve this one.

**Comment by i-oden** [comment]
> Everything seems good to me. Probably don't yet merge though, before Ina's approval. I think she wanted to explicitly approve this one.

@rv0lt Correct. In the future, don't merge PRs that have a direct affect on the users unless I've explicitly said go ahead. I'll review this after merge this time though and if we'll see if there's a new PR needed.

**Comment by i-oden** [comment]
Also, when we have these kinds of tasks that require both an API and a CLI part, as I mentioned I would like us to have a PR approved for both before merging. Otherwise, we may notice that we need changes in the API again when we're making the CLI part.

**Comment by rv0lt** [comment]
> > Everything seems good to me. Probably don't yet merge though, before Ina's approval. I think she wanted to explicitly approve this one.
> 
> @rv0lt Correct. In the future, don't merge PRs that have a direct affect on the users unless I've explicitly said go ahead. I'll review this after merge this time though and if we'll see if there's a new PR needed.

Sorry, I didn't saw the comment from Valentin, I wanted to close them because there were quite a bunch of tasks before the finish of the sprint tomorrow so I got a bit anxious and only saw the notification

File: dds_web/api/superadmin_only.py [hunk 2]
```diff
@@ -163,6 +163,20 @@ def post(self):
         if not motd_obj or not motd_obj.active:
             raise ddserr.DDSArgumentError(message=f"There is no active MOTD with ID '{motd_id}'.")
 
+        # check if sent to unit personnel only or all users
+        unit_personnel_only: bool = request_json.get("unit_personnel_only", False)
+        if not isinstance(unit_personnel_only, bool):
+            raise ddserr.DDSArgumentError(
+                message="The 'unit_personnel_only' argument must be a boolean."
+            )
```
Comment (i-oden) [inline dds_web/api/superadmin_only.py:171]: As mentioned in the CLI PR, `unit_personnel_only` sounds like we're only sending to `Unit Personnel`, and not also `Unit Admin`s. Should be changed to `unit users` in this case.

File: dds_web/api/superadmin_only.py [hunk 3]
```diff
@@ -163,6 +163,20 @@ def post(self):
...
+        if unit_personnel_only:
+            unit_usernames = db.session.query(models.UnitUser.username)
+            users_to_send = db.session.query(models.User).filter(
+                models.User.username.in_(unit_usernames)
+            )
```
Comment (i-oden) [inline dds_web/api/superadmin_only.py:176]: This is unnecessary, all that's needed here is this: 
```python
            users_to_send = db.session.query(models.UnitUser)
```

**Review by i-oden — COMMENTED** [review]
1. Adding a new card to the product backlog for these changes along with the ones in https://github.com/ScilifelabDataCentre/dds_cli/pull/717, but also, there's no description of the changes in this PR (it just says "TODO"). Update this.
2. The test coverage was not 100%, right? Can you add info to the PR about which lines were not covered and the reason for it? Thanks.
3. I'm confused, why are we adding new tests to multiple locations? Both `tests/api` and `tests/tests_v3/api`?

**Comment by rv0lt** [comment]
> 1. Adding a new card to the product backlog for these changes along with the ones in [Modify the command dds motd send to allow the option to send only to unit personnel dds_cli#717](https://github.com/ScilifelabDataCentre/dds_cli/pull/717), but also, there's no description of the changes in this PR (it just says "TODO"). Update this.
> 
>     2. The test coverage was not 100%, right? Can you add info to the PR about which lines were not covered and the reason for it? Thanks.
> 
>     3. I'm confused, why are we adding new tests to multiple locations? Both `tests/api` and `tests/tests_v3/api`?

2. Test coverage is full for this PR. It is not 100% in the whole repository, but there is an increase of **hits** and no change in **misses**

https://app.codecov.io/gh/ScilifelabDataCentre/dds_web/pull/1552?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ScilifelabDataCentre

image


3. That is part of the new API version that is located in /api/v3 to be OpenAPI compliant. To achieve that, some requests had to change where do they receive data (From requests body to parameters) and therefore requiere a different set of tests. You can check out the two version endpoints and try them in

- https://testing.delivery.scilifelab.se/api/documentation/v3/
- https://testing.delivery.scilifelab.se/api/documentation/v1/

Also, in order to have documented in Swagger, you need to be compliant as well, as it doesn't let you otherwise, and thus, incompatible methods are marked as deprecated in the v1 version

**Comment by i-oden** [comment]
> > 1. Adding a new card to the product backlog for these changes along with the ones in [Modify the command dds motd send to allow the option to send only to unit personnel dds_cli#717](https://github.com/ScilifelabDataCentre/dds_cli/pull/717), but also, there's no description of the changes in this PR (it just says "TODO"). Update this.
> >    
> >    2. The test coverage was not 100%, right? Can you add info to the PR about which lines were not covered and the reason for it? Thanks.
> >    3. I'm confused, why are we adding new tests to multiple locations? Both `tests/api` and `tests/tests_v3/api`?
> 
>     2. Test coverage is full for this PR. It is not 100% in the whole repository, but there is an increase of **hits** and no change in **misses**
> 
> 
> https://app.codecov.io/gh/ScilifelabDataCentre/dds_web/pull/1552?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ScilifelabDataCentre
> 
> image
> 
>     3. That is part of the new API version that is located in /api/v3 to be OpenAPI compliant. To achieve that, some requests had to change where do they receive data (From requests body to parameters) and therefore requiere a different set of tests. You can check out the two version endpoints and try them in
> 
> 
>     * https://testing.delivery.scilifelab.se/api/documentation/v3/
> 
>     * https://testing.delivery.scilifelab.se/api/documentation/v1/
> 
> 
> Also, in order to have documented in Swagger, you need to be compliant as well, as it doesn't let you otherwise, and thus, incompatible methods are marked as deprecated in the v1 version

Regarding the 100% coverage, I mean for the changes made in the PR. Are the changes covered 100% and if not why? For documentation purposes and to make sure that we know if there's something that needs to be fixed or added. 

Regarding Swagger -- I did see that we have two different documentations, and that's good. Good that we're making it all OpenAPI compliant. However, I still don't understand how this is affects the tests? Am I missing something? In my mind, we'd just update the tests, and potentially put those tests in the v3 location? Not have duplicates?

**Comment by rv0lt** [comment]
2.
Yes, that's what I mean. It says _All modified and coverable lines are covered by tests_ And project coverage will go up from 92.50 to 92.52

3.

Ahh, I see the question now. 
Yes, so the v3 test files are duplicates of the files in the root for two reasons mainly: One is that, in the future, when the v1 is removed, then to replace the tests is would be just replacing the file. And the other is that, even though the tests are the same, in every case it is being called with a different endpoint version. And maybe, there can be some change sometime that only affects one of them. So it's a matter of having that covered

**Comment by i-oden** [comment]
> Yes, that's what I mean. It says _All modified and coverable lines are covered by tests_ And project coverage will go up from 92.50 to 92.52
> 
> 
> Ahh, I see the question now. Yes, so the v3 test files are duplicates of the files in the root for two reasons mainly: One is that, in the future, when the v1 is removed, then to replace the tests is would be just replacing the file. And the other is that, even though the tests are the same, in every case it is being called with a different endpoint version. And maybe, there can be some change sometime that only affects one of them. So it's a matter of having that covered

2. Ah sorry, I missed that somehow. Great. 
3. Okey, but if the files are duplicates, then why can't we just have the v1? Then we wouldn't need to replace any files, we're just updating the tests as usual. Or is swagger somehow dependent on / requires a specific, unique directory of tests? I don't think I fully understand, I think I need someone to explain this to me on Zoom and perhaps share screen at the same time, probably easier than having this discussion here.
